src/frustum.rs

```diff
@@ -48,14 +49,38 @@ Frustum<S> {
         }
     }
 
-    /// Extracts frustum planes from a projection matrix
-    pub fn from_matrix4(mat: Matrix4<S>) -> Frustum<S> {
-        Frustum::new(Plane::from_vector4(mat.row(3).add_v(&mat.row(0)).normalize()),
-                     Plane::from_vector4(mat.row(3).sub_v(&mat.row(0)).normalize()),
-                     Plane::from_vector4(mat.row(3).add_v(&mat.row(1)).normalize()),
-                     Plane::from_vector4(mat.row(3).sub_v(&mat.row(1)).normalize()),
-                     Plane::from_vector4(mat.row(3).add_v(&mat.row(2)).normalize()),
-                     Plane::from_vector4(mat.row(3).sub_v(&mat.row(2)).normalize()))
+    /// Extract frustum planes from a projection matrix.
+    pub fn from_matrix4(mat: Matrix4<S>) -> Option<Frustum<S>> {
+        Some(Frustum::new(
+            match Plane::from_vector4_alt(mat.row(3).add_v(&mat.row(0))).normalize()
+                { Some(p) => p, None => return None },
+            match Plane::from_vector4_alt(mat.row(3).sub_v(&mat.row(0))).normalize()
+                { Some(p) => p, None => return None },
+            match Plane::from_vector4_alt(mat.row(3).add_v(&mat.row(1))).normalize()
+                { Some(p) => p, None => return None },
+            match Plane::from_vector4_alt(mat.row(3).sub_v(&mat.row(1))).normalize()
+                { Some(p) => p, None => return None },
+            match Plane::from_vector4_alt(mat.row(3).add_v(&mat.row(2))).normalize()
+                { Some(p) => p, None => return None },
+            match Plane::from_vector4_alt(mat.row(3).sub_v(&mat.row(2))).normalize()
+                { Some(p) => p, None => return None }
+        ))
+    }
+
+    /// Find the spatial relation of a bound inside this frustum.
+    pub fn contains<B: Bound<S>>(&self, bound: &B) -> Relation {
+        [&self.left, &self.right, &self.top, &self.bottom, &self.near, &self.far]
+            .iter().fold(Relation::In, |cur, p| {
+            let r = bound.relate_plane(p);
+            // if we see Cross, then the result is Cross
+            // if we see In, then we keep the old result
+            // otherwise, take the current result
+            if cur == Relation::Cross || r == Relation::In {
```

Comment:
`cur == Out && r == In` would result in `Out`, since it's the first arm of the condition. You are correct about the second case though, `Cross` shouldn't overwrite `Out`, and I need to fix the condition.